### PR TITLE
Core Data integration

### PR DESCRIPTION
**Pull request by heyjaywilson**
Progress / Closed on issue #

## What it Does

TLDR: Introduces Core Data as the main data store 💽

> 📙 I left some more explanations in comments in the code cause that's easier for future maintainers to see.

### Adds Core Data for Tweet and Category CRUD

Core Data has been added and the supporting files can be found in the `CoreData` folder.

There are two entities.

- `CategoryEntity` holds the categories and has a one to many relationship to the `TweetEntity`. 
- `TweetEntity` holds the tweets and has a many to one relationship to the `CategoryEntity`.

`CoreData/StorageProvider.swift` is used to initialize the `NSPersistentContainer`. It has `preview` variable which is used for the SwiftUI canvas/preview and a `shared` variable which is used all other times.

#### How a migration is performed:

1. On app launch, a check is done if the migration needs to happen.
2. All categories are fetched from AWS Amplify
3. For each category:
    1. Translate the AWS Category to a `CoreDataEntity`
    2. Fetch the AWS Tweets from the AWS Category
    3. For each tweet, translate it to a `TweetEntity` and add it to the `CategoryEntity` tweets
4. Save to core data  

#### How is a Migration determined:
The following conditions must both be true for a migration to run.
1. `storageProvider.getAllCategories().isEmpty` is true
5. The user default `migrationToCoreDataRan` is true

## How I Tested

1. Run the application from `dev` branch, create a category and save some tweets
6. Run the application from `jay/bugfix-data` and see the categories look the same (migration ran)
7. Save some more tweets and the tweets persist through app load
8. Edit the category name and see it stays the same through app loads

### CODE REVIEW

File: brain-marks/Services/MigrationService.swift [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            switch result {
+            case .success(let categories):
+                self.awsCategories = categories
+            case .failure(let error):
+                print("❌ MigrationService.getCategories(): Error: \(error)")
+            }
+        })
+    }
+
+    private func makeCategoryCoreDataCompatible(category: AWSCategory) -> CategoryEntity {
```
Comment (heyjaywilson) [inline brain-marks/Services/MigrationService.swift]: Naming is hard. I'm down to rename this but I'm not too sure to what.

File: brain-marks/Services/MigrationService.swift [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        tweetCat.id = UUID()
+        tweetCat.dateCreated = Date()
+        tweetCat.dateModified = Date()
+        tweetCat.imageName = category.imageName ?? "folder"
+        tweetCat.name = category.name
+
+        return tweetCat
+    }
+
+    private func makeTweetCoreDataCompatible(tweet: AWSTweet) -> TweetEntity {
```
Comment (heyjaywilson) [inline brain-marks/Services/MigrationService.swift]: Naming is hard. I'm down to rename this but I'm not too sure to what.

File: brain-marks/Categories/Views/CategoryList.swift [hunk 3]
```diff
@@ -82,9 +78,12 @@ struct CategoryList: View {
     
     var categories: some View {
         List {
-            ForEach(viewModel.categories) { category in
+            ForEach(viewModel.categories, id: \.self) { category in
                 NavigationLink(destination: TweetList(category: category)) {
-                    CategoryRow(category: category)
+                    CategoryRow(
```
Comment (heyjaywilson) [inline brain-marks/Categories/Views/CategoryList.swift]: This needed to change to take in both the name and image, since those change but the ID doesn't. Since the ID wasn't changing this wasn't updating.

File: brain-marks/ContentView.swift [hunk 4]
```diff
@@ -8,10 +8,17 @@
 import SwiftUI
 
 struct ContentView: View {
+    let migrationService: MigrationService = MigrationService()
```
Comment (mikaelacaron) [inline brain-marks/ContentView.swift]:
```suggestion
    let migrationService = MigrationService()
```

File: brain-marks/ContentView.swift [hunk 5]
```diff
@@ -8,10 +8,17 @@
...
+
     @State private var showAddSheet = false
 
     var body: some View {
         CategoryList()
+            .onAppear {
+                if migrationService.checkIfMigrationShouldRun() {
```
Comment (mikaelacaron) [inline brain-marks/ContentView.swift:19]: Let's add a telemetry signal that a user's data was migrated

`migrationFromAmplifyToCDPerformed`

File: brain-marks/Services/MigrationService.swift [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+//  brain-marks
+//
+//  Created by Jay on 12/13/22.
+//
+
+import CoreData
+import Foundation
+
+/// Controls the migration from Amplify to CoreData
+class MigrationService {
```
Comment (mikaelacaron) [inline brain-marks/Services/MigrationService.swift]:
```suggestion
final class MigrationService {
```

Because this doesn't need to be inherited by anything

File: brain-marks/Services/MigrationService.swift [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            do {
+                try managedObjectContext.save()
+            } catch {
+                print("❌ MigrationService.performMigration() Error: \(error)")
+            }
+        }
+        UserDefaults.standard.set(true, forKey: "migrationToCoreDataRan")
+    }
+
+    public func checkIfMigrationShouldRun() -> Bool {
```
Comment (mikaelacaron) [inline brain-marks/Services/MigrationService.swift]:
```suggestion
    func checkIfMigrationShouldRun() -> Bool {
```

This doesn't need to be public, does it?

File: brain-marks/Services/MigrationService.swift [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    private var managedObjectContext: NSManagedObjectContext
+
+    init() {
+        self.managedObjectContext = storageProvider.context
+    }
+
+    func performMigration() {
+        /// 1. Fetch categories
+        getCategories()
```
Comment (mikaelacaron) [inline brain-marks/Services/MigrationService.swift]: technically because `getCategories` is called here and sets the variable `awsCategories` is it possible for the function to still be fetching data, before moving onto the next step? (iterating through `awsCategories`)

I want to say maybe? and because of this, let's not use this function `getCategories` and instead use it directly in `performMigration` even if the code isn't that clean. This also makes sure that the UserDefault bool doesn't get set as true, even if this `getCategories` failed for some reason

File: brain-marks/Services/MigrationService.swift (hunk 1 above)
Comment (mikaelacaron) [reply brain-marks/Services/MigrationService.swift]: Let's do `convertAmplifyCategoryToCoreDataCategory`

File: brain-marks/Services/MigrationService.swift (hunk 2 above)
Comment (mikaelacaron) [reply brain-marks/Services/MigrationService.swift]: `convertAmplifyTweetToCoreDataTweet`

File: brain-marks/brain_marksApp.swift [hunk 9]
```diff
@@ -12,15 +12,17 @@ import TelemetryClient
 
 @main
 struct brain_marksApp: App {
-    
+    @AppStorage("migrationToCoreDataRan")
```
Comment (mikaelacaron) [inline brain-marks/brain_marksApp.swift:15]: what does adding this variable here do, when it isn't used?

File: brain-marks/CoreData/StorageProvider.swift [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                category.addToTweets(tweet)
+            }
+        }
+        return controller
+    }()
+
+    init(inMemory: Bool = false) {
+        container = NSPersistentContainer(name: "BrainMarks")
+
+        if inMemory {
```
Comment (mikaelacaron) [inline brain-marks/CoreData/StorageProvider.swift:49]: what does this do?
Also considering this is a singleton, the `init` is supposed to be private, because you aren't supposed to have multiple instances of it at a time. Would this affect what the init does?

File: brain-marks/CoreData/StorageProvider.swift [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+//  brain-marks
+//
+//  Created by Jay on 12/13/22.
+//
+
+import Foundation
+import CoreData
+import TelemetryClient
+
+class StorageProvider {
```
Comment (mikaelacaron) [inline brain-marks/CoreData/StorageProvider.swift]: Could this be `final`?

File: brain-marks/CoreData/StoreageProvider+Category.swift [hunk 12]
```diff
@@ -0,0 +1,33 @@
+//
+//  StoreageProvider+Category.swift
```
Comment (mikaelacaron) [inline brain-marks/CoreData/StoreageProvider+Category.swift]: filename is spelled wrong

File: brain-marks/CoreData/StoreageProvider+Category.swift [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    func getAllCategories() -> [CategoryEntity] {
+        let fetchRequest: NSFetchRequest<CategoryEntity> = CategoryEntity.fetchRequest()
+        fetchRequest.sortDescriptors = [NSSortDescriptor(key: "dateModified", ascending: false)]
+
+        do {
+            return try context.fetch(fetchRequest)
+        } catch {
+            print("StorageProvider.getAllCategories(:): Error \(error)")
```
Comment (mikaelacaron) [inline brain-marks/CoreData/StoreageProvider+Category.swift]:
```suggestion
            print("❌ StorageProvider.getAllCategories(:): Error \(error)")
```

File: brain-marks/Categories/Views/CategoryRow.swift [hunk 14]
```diff
@@ -8,15 +8,16 @@
 import SwiftUI
 
 struct CategoryRow: View {
-    let category: AWSCategory
+    let categoryName: String
```
Comment (mikaelacaron) [inline brain-marks/Categories/Views/CategoryRow.swift]: I'd rather keep this as passing a single category object, rather than 2 properties, because it's possible for this to grow

**Review by mikaelacaron — CHANGES_REQUESTED** [review]
We'll chat in Discord 👍

File: brain-marks/brain_marksApp.swift (hunk 9 above)
Comment (heyjaywilson) [reply brain-marks/brain_marksApp.swift:15]: the purpose of this being here is to run before the migration service does on first launch. It'll default to false on the first app run since being introduced. It's then set to true to not run the migration again.

File: brain-marks/CoreData/StorageProvider.swift (hunk 10 above)
Comment (heyjaywilson) [reply brain-marks/CoreData/StorageProvider.swift:49]: I'll make the init private, but this tells the app which provider to look at for tests and such.

File: brain-marks/Categories/Views/CategoryRow.swift (hunk 14 above)
Comment (heyjaywilson) [reply brain-marks/Categories/Views/CategoryRow.swift]: Done ✅

File: brain-marks/CoreData/StoreageProvider+Category.swift (hunk 13 above)
Comment (heyjaywilson) [reply brain-marks/CoreData/StoreageProvider+Category.swift]: Done ✅

File: brain-marks/CoreData/StoreageProvider+Category.swift (hunk 12 above)
Comment (heyjaywilson) [reply brain-marks/CoreData/StoreageProvider+Category.swift]: 🤦 Done ✅

File: brain-marks/CoreData/StorageProvider.swift (hunk 11 above)
Comment (heyjaywilson) [reply brain-marks/CoreData/StorageProvider.swift]: can do Done ✅

File: brain-marks/Services/MigrationService.swift (hunk 7 above)
Comment (heyjaywilson) [reply brain-marks/Services/MigrationService.swift]: Nope it doesn't

File: brain-marks/Services/MigrationService.swift (hunk 6 above)
Comment (heyjaywilson) [reply brain-marks/Services/MigrationService.swift]: Done ✅

File: brain-marks/ContentView.swift (hunk 5 above)
Comment (heyjaywilson) [reply brain-marks/ContentView.swift:19]: Done ✅

File: brain-marks/ContentView.swift (hunk 4 above)
Comment (heyjaywilson) [reply brain-marks/ContentView.swift]: Done ✅

File: brain-marks/Services/MigrationService.swift (hunk 2 above)
Comment (heyjaywilson) [reply brain-marks/Services/MigrationService.swift]: Done ✅  Thanks for a better name

File: brain-marks/Services/MigrationService.swift (hunk 1 above)
Comment (heyjaywilson) [reply brain-marks/Services/MigrationService.swift]: Dane! Thanks for a better name!

File: brain-marks/Services/MigrationService.swift (hunk 8 above)
Comment (heyjaywilson) [reply brain-marks/Services/MigrationService.swift]: Am I understanding this right, move the `amplifyDataStore.fetchCategories` into `performMigration`?

I think that makes a bit more sense also

File: brain-marks/CoreData/StorageProvider.swift (hunk 10 above)
Comment (mikaelacaron) [reply brain-marks/CoreData/StorageProvider.swift:49]: Are there any unit tests for this? I didn't see any @heyjaywilson
